### PR TITLE
FIX: DB.json().toBean(target) can update existing lists

### PR DESCRIPTION
**Pull request by rPraml**

### CODE REVIEW

File: ebean-core/src/main/java/io/ebeaninternal/api/json/SpiJsonReader.java [hunk 1]
```diff
@@ -34,5 +34,5 @@ public interface SpiJsonReader {
 
   Object readValueUsingObjectMapper(Class<?> propertyType) throws IOException;
 
-  boolean intercept();
+  boolean update();
```
Comment (rPraml) [inline ebean-core/src/main/java/io/ebeaninternal/api/json/SpiJsonReader.java:37]: I've changed `intercept` to `update` as I think this is a better name now

File: ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanDescriptorJsonHelp.java [hunk 2]
```diff
@@ -133,7 +133,12 @@ private T jsonReadProperties(SpiJsonReader readJson, EntityBean bean, String pat
         String key = parser.getCurrentName();
         BeanProperty p = desc.beanProperty(key);
         if (p != null) {
-          p.jsonRead(readJson, bean);
+          if (p.isVersion() && readJson.update() ) {
+            // skip version prop during updte
+            p.jsonRead(readJson);
```
Comment (rPraml) [inline ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanDescriptorJsonHelp.java:138]: when we update existing beans, we will skip the version with a dummy read

File: ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanProperty.java [hunk 3]
```diff
@@ -1416,15 +1416,20 @@ private void jsonWriteScalar(SpiJsonWriter writeJson, Object value) throws IOExc
...
+    if (jsonDeserialize) {
+      if (ctx.update()) {
+        setValueIntercept(bean, objValue);
+      } else {
+        setValue(bean, objValue);
+      }
+    }
+  }
+
+  public Object jsonRead(SpiJsonReader ctx) throws IOException {
```
Comment (rPraml) [inline ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanProperty.java:1429]: I've split the method to support reading a property without setting it to the bean

File: ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanPropertyAssocMany.java [hunk 4]
```diff
@@ -1022,15 +1025,16 @@ private void jsonWriteCollection(SpiJsonWriter ctx, String name, Object value, b
...
-  public Object jsonReadCollection(SpiJsonReader readJson, EntityBean parentBean) throws IOException {
+  public Object jsonReadCollection(SpiJsonReader readJson, EntityBean parentBean, Object targets) throws IOException {
     if (elementDescriptor != null && elementDescriptor.isJsonReadCollection()) {
       return elementDescriptor.jsonReadCollection(readJson, parentBean);
     }
     BeanCollection<?> collection = createEmpty(parentBean);
     BeanCollectionAdd add = beanCollectionAdd(collection);
+    Map<Object, T> existingBeans = extractBeans(targets);
     do {
-      // CHECKME: Update existing list entry here?
```
Comment (rPraml) [inline ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanPropertyAssocMany.java:1032]: YES: we have to update existing list/map/set here

File: ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanPropertyAssocOne.java [hunk 5]
```diff
@@ -811,16 +811,22 @@ public void jsonWrite(SpiJsonWriter writeJson, EntityBean bean) throws IOExcepti
   @Override
   public void jsonRead(SpiJsonReader readJson, EntityBean bean) throws IOException {
     if (jsonDeserialize && targetDescriptor != null) {
-      T target = (T) value(bean);
+      // CHECKME: may we skip reading the object from the json stream?
```
Comment (rPraml) [inline ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanPropertyAssocOne.java:814]: I have not changed the behaviour here, but "BeanProperty" will always read the value from the json stream (and discards it, when `jsonDeserialize==false`)
Here we do not consume the bean from the jsonReader

File: ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanDescriptorJsonHelp.java [hunk 6]
```diff
@@ -133,7 +133,12 @@ private T jsonReadProperties(SpiJsonReader readJson, EntityBean bean, String pat
         String key = parser.getCurrentName();
         BeanProperty p = desc.beanProperty(key);
         if (p != null) {
-          p.jsonRead(readJson, bean);
+          if (p.isVersion() && readJson.update() ) {
+            // skip version prop during updte
```
Comment (nPraml) [inline ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanDescriptorJsonHelp.java:137]: `// skip version prop during update`

File: ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanProperty.java [hunk 7]
```diff
@@ -20,9 +20,9 @@
 import io.ebeaninternal.api.SpiQuery;
 import io.ebeaninternal.api.json.SpiJsonReader;
 import io.ebeaninternal.api.json.SpiJsonWriter;
+import io.ebeaninternal.server.bind.DataBind;
 import io.ebeaninternal.server.core.EncryptAlias;
 import io.ebeaninternal.server.core.InternString;
-import io.ebeaninternal.server.bind.DataBind;
```
Comment (nPraml) [inline ebean-core/src/main/java/io/ebeaninternal/server/deploy/BeanProperty.java:25]: ich glaube dass die Reihenfolge gepasst hat

File: ebean-test/src/test/java/org/tests/text/json/TestJsonBeanDescriptorParse.java [hunk 8]
```diff
@@ -32,7 +37,20 @@ void setup() {
...
+    ContactNote note = new ContactNote("Drinks", "Order 100l beer");
+    note.setId(890);
+    alfred.getNotes().add(note);
+    customer.getContacts().add(alfred);
+    Contact anton = new Contact("Anton", "P");
+    anton.setId(790);
+    anton.getNotes().add(new ContactNote("Equipment", "Organize barbecue"));
+    customer.getContacts().add(anton);
     DB.save(customer);
+    System.out.println(DB.json().toJson(customer));
```
Comment (nPraml) [inline ebean-test/src/test/java/org/tests/text/json/TestJsonBeanDescriptorParse.java:52]: soll man das später löschen oder braucht man es beim Test als "Prüfung"?

**Review by nPraml — APPROVED** [review]

**Pull request merged**